### PR TITLE
Add SVD pipeline

### PR DESCRIPTION
**Pull request by dsocek**
# What does this PR do?
- implements Stable Video Diffusion (SVD) pipeline for Gaudi
- adds example SVD inference script and README
- adds SVD related CI tests

### CODE REVIEW

File: optimum/habana/diffusers/models/unet_spatio_temporal_condition.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            The additional time ids with shape `(batch, num_additional_ids)`. These are encoded with sinusoidal
+            embeddings and added to the time embeddings.
+        return_dict (`bool`, *optional*, defaults to `True`):
+            Whether or not to return a [`~models.unet_slatio_temporal.UNetSpatioTemporalConditionOutput`] instead of a plain
+            tuple.
+    Returns:
+        [`~models.unet_slatio_temporal.UNetSpatioTemporalConditionOutput`] or `tuple`:
+            If `return_dict` is True, an [`~models.unet_slatio_temporal.UNetSpatioTemporalConditionOutput`] is returned, otherwise
+            a `tuple` is returned where the first element is the sample tensor.
+    """
```
Comment (yeonsily) [inline optimum/habana/diffusers/models/unet_spatio_temporal_condition.py:39]: Can you please put comment that where the original code came from and what hpu specific change added here?

File: optimum/habana/diffusers/models/unet_spatio_temporal_condition.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    # Repeat the embeddings num_video_frames times
+    # emb: [batch, channels] -> [batch * frames, channels]
+    emb = emb.repeat_interleave(num_frames, dim=0)
+    # encoder_hidden_states: [batch, 1, channels] -> [batch * frames, 1, channels]
+    encoder_hidden_states = encoder_hidden_states.repeat_interleave(num_frames, dim=0)
+
+    # 2. pre-process
+    import habana_frameworks.torch.hpu as hthpu
+
+    # Workaround for SynapseAI 1.11 for Torch Autocast
```
Comment (yeonsily) [inline optimum/habana/diffusers/models/unet_spatio_temporal_condition.py:85]: Can you please give some details what's the autocast issue here?

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+            self,
+            use_habana,
+            use_hpu_graphs,
+            gaudi_config,
+            bf16_full_eval,
+        )
+
+        # Workaround for Synapse 1.11 for full bf16
+        if bf16_full_eval:
+            unet.conv_in.float()
```
Comment (libinta) [inline optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:106]: Can you check why do we still need this as float?

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,762 @@
...
+
+        passed_add_embed_dim = self.unet.config.addition_time_embed_dim * len(add_time_ids)
+        expected_add_embed_dim = self.unet.add_embedding.linear_1.in_features
+
+        if expected_add_embed_dim != passed_add_embed_dim:
+            raise ValueError(
+                f"Model expects an added time embedding vector of length {expected_add_embed_dim}, but a vector of {passed_add_embed_dim} was created. The model has an incorrect config. Please check `unet.config.time_embedding_type` and `text_encoder_2.config.projection_dim`."
+            )
+
+        add_time_ids = torch.tensor([add_time_ids], dtype=dtype)
```
Comment (libinta) [inline optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:204]: this tensor is on cpu, please change to hpu

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py (hunk 3 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:106]: @libinta I checked on 1.14 and 1.15 (I did't check 1.12 and 1.13), and we do not need this WA anymore.   I forced `bf16_full_eval` to `True` and removed WA and pipeline worked OK.   I will remove in amended revision

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py (hunk 4 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:204]: @libinta Thank you for catching this, I will move it to HPU in the PR amendment

File: optimum/habana/diffusers/models/unet_spatio_temporal_condition.py (hunk 2 above)
Comment (dsocek) [reply optimum/habana/diffusers/models/unet_spatio_temporal_condition.py:85]: @yeonsily I copied same WA from https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/models/unet_2d_condition.py#L204

Not sure if its still needed for 1.14,1.15 version (probably not), I will check

File: optimum/habana/diffusers/models/unet_spatio_temporal_condition.py (hunk 1 above)
Comment (dsocek) [reply optimum/habana/diffusers/models/unet_spatio_temporal_condition.py:39]: I will add comment on source and HPU changes

File: optimum/habana/diffusers/models/unet_spatio_temporal_condition.py (hunk 2 above)
Comment (dsocek) [reply optimum/habana/diffusers/models/unet_spatio_temporal_condition.py]: 1.11 WA not needed for 1.15.  I removed modified forward unet call from PR as its obsolete

File: optimum/habana/diffusers/models/unet_spatio_temporal_condition.py (hunk 1 above)
Comment (dsocek) [reply optimum/habana/diffusers/models/unet_spatio_temporal_condition.py]: I removed modified forward unet call from PR as its obsolete (WA for 1.11 not needed anymore)

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py (hunk 3 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py]: Fixed

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py (hunk 4 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py]: Done

**Comment by dsocek** [comment]
Note:  Quality of generated video is really poor with `classifier_free_guidance` enabled.   I override it to `False` and quality looks good.

File: examples/stable-video-diffusion/README.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
+<!---
+Copyright 2024 The HuggingFace Team. All rights reserved.
```
Comment (libinta) [inline examples/stable-video-diffusion/README.md:2]: can we move the full thing to stable diffusion folder?

File: examples/stable-video-diffusion/image_to_video_generation.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        default="stabilityai/stable-video-diffusion-img2vid-xt",
+        type=str,
+        help="Path to pre-trained model",
+    )
+
+    # Pipeline arguments
+    parser.add_argument(
+        "--image_path",
+        type=str,
+        default="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/svd/rocket.png",
```
Comment (libinta) [inline examples/stable-video-diffusion/image_to_video_generation.py:57]: should we put None here?

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,763 @@
...
+)
+from diffusers.schedulers import EulerDiscreteScheduler
+from diffusers.utils import BaseOutput, logging
+from diffusers.utils.torch_utils import is_compiled_module, randn_tensor
+from transformers import CLIPImageProcessor, CLIPVisionModelWithProjection
+
+from ....transformers.gaudi_configuration import GaudiConfig
+from ....utils import speed_metrics
+from ..pipeline_utils import GaudiDiffusionPipeline
+
```
Comment (libinta) [inline optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:39]: can you add a link for the source of the file?

Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:39]: Done

File: examples/stable-video-diffusion/image_to_video_generation.py (hunk 6 above)
Comment (dsocek) [reply examples/stable-video-diffusion/image_to_video_generation.py]: Done

File: examples/stable-video-diffusion/README.md (hunk 5 above)
Comment (dsocek) [reply examples/stable-video-diffusion/README.md]: Done

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+                image_embeddings_batch = image_embeddings_batches[0]
+                image_embeddings_batches = torch.roll(image_embeddings_batches, shifts=-1, dims=0)
+                added_time_ids_batch = added_time_ids_batches[0]
+                added_time_ids_batches = torch.roll(added_time_ids_batches, shifts=-1, dims=0)
+
+                for i in self.progress_bar(range(num_inference_steps)):
+                    timestep = timesteps[0]
+                    timesteps = torch.roll(timesteps, shifts=-1, dims=0)
+
+                    capture = True if self.use_hpu_graphs and j == 0 and i < 2 else False
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py:685]: There is no HPU graphs at the moment no?

File: examples/stable-diffusion/image_to_video_generation.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        args.model_name_or_path,
+        **kwargs,
+    )
+
+    # Set seed before running the model
+    set_seed(args.seed)
+
+    # Load input image(s)
+    input = []
+    print("Input image(s):")
```
Comment (regisss) [inline examples/stable-diffusion/image_to_video_generation.py:201]:
```suggestion
    logger.info("Input image(s):")
```

File: examples/stable-diffusion/image_to_video_generation.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    if isinstance(args.image_path, str):
+        args.image_path = [args.image_path]
+    for image_path in args.image_path:
+        image = load_image(image_path)
+        image = image.resize((args.height, args.width))
+        input.append(image)
+        print(image_path)
```
Comment (regisss) [inline examples/stable-diffusion/image_to_video_generation.py:208]:
```suggestion
        logger.info(image_path)
```

File: examples/stable-diffusion/README.md [hunk 11]
```diff
@@ -276,3 +276,47 @@ python text_to_image_generation.py \
...
 ```
+
+# Stable Video Diffusion Examples
+
+Stable Video Diffusion (SVD) was unveiled in [Stable Video Diffusion Announcement](https://stability.ai/news/stable-video-diffusion-open-ai-video-model)
+by the Stability AI team. Stable Video Diffusion XT version (SVD-XT) is tuned to generate 25 frames of video from a single image.
+
+## Image-to-video Generation
+
+Script `image_to_video_generation.py` showcases how to perform image-to-video generation using Stable Video Diffusion on Habana Gaudi.
```
Comment (regisss) [inline examples/stable-diffusion/README.md:287]:
```suggestion
Script `image_to_video_generation.py` showcases how to perform image-to-video generation using Stable Video Diffusion on Intel Gaudi.
```

**Review by regisss — COMMENTED** [review]
There is no support for HPU graphs as of now?

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py [hunk 12]
```diff
@@ -0,0 +1,767 @@
...
+        unet ([`UNetSpatioTemporalConditionModel`]):
+            A `UNetSpatioTemporalConditionModel` to denoise the encoded image latents.
+        scheduler ([`EulerDiscreteScheduler`]):
+            A scheduler to be used in combination with `unet` to denoise the encoded image latents.
+        feature_extractor ([`~transformers.CLIPImageProcessor`]):
+            A `CLIPImageProcessor` to extract features from generated images.
+    """
+
+    model_cpu_offload_seq = "image_encoder->unet->vae"
+    _callback_tensor_inputs = ["latents"]
```
[the remark this answers is not in the document]
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py]: Done

File: optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py (hunk 8 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_video_diffusion/pipeline_stable_video_diffusion.py]: No seems graph capture is failing at the moment

File: examples/stable-diffusion/image_to_video_generation.py (hunk 9 above)
Comment (dsocek) [reply examples/stable-diffusion/image_to_video_generation.py]: Fixed

File: examples/stable-diffusion/image_to_video_generation.py (hunk 10 above)
Comment (dsocek) [reply examples/stable-diffusion/image_to_video_generation.py]: Fixed

File: examples/stable-diffusion/README.md (hunk 11 above)
Comment (dsocek) [reply examples/stable-diffusion/README.md]: Fixed

**Comment by dsocek** [comment]
@regisss  Thanks for review, yes for now HPU graph capture disabled.  Fixed all other indicated issues in the amendment.

**Comment by dsocek** [comment]
@libinta @regisss  Replaced CPU-based `randn_tensor` call with HPU-based (tested and output video looks good)

**Comment by regisss** [comment]
@dsocek LGTM! Could you also add a test in `tests/test_diffusers.py` to test the throughput and that the example works please?

**Comment by dsocek** [comment]
> @dsocek LGTM! Could you also add a test in `tests/test_diffusers.py` to test the throughput and that the example works please?

@regisss added SVD test

**Review by regisss — APPROVED** [review]
